Shopify wishlist implementation

chemmanouilidis opened the pull request:
Wishlist implementation for the Shopify provider. You can Add, Remove products to your customer wishlist as well as their variants. Also Added the possibility to add a specific product variant to your wishlist if its not available for sale (due to no stock etc)...

1) changed shopify fetch and graphQL fetch types and also the fetcher functions to include useAdminApiFlag in order to seperate graphql storefront API operations from Admin API ones, 
2) generated shopify admin api schema(documentation for code generation is outdated and not working as per readme instructions - yarn generate:shopify script not present), 
3) changed RemoveItemHook of commerce/types/wishlist, 
4) created shopify wishlist api routes (addItem, removeItem, getWishlist), 
5) added shopify api operation getCustomerWishlist, 
6) updated index files of api/operations, utils and api/wishlist, 
7) created shopify wishlist types ,
8) created getCustomerId function as a util in shopify ,
9) added customerUpdate mutation (Admin API operation),  customer query (Admin API operation), 
10) added getCustomerId query,
11) modified commerce wishlist endpoint to extract item variable from req/body, 
12) added useAdminApi flag to commerce fetcherOptions, 
13) updated wishlist api constants, 
14) modified shopify fetcher, 
15) added wishlist handlers to shopify provider ,
16) Polished frontend UI so that you can now correctly add product to wishlist if variant is not available for sale

17) modified .env.template files to correctly refer that you also need shopify admin access token

dominiksipowicz commented on packages/commerce/src/api/endpoints/wishlist.ts:
```diff
@@ -34,13 +34,13 @@ const wishlistEndpoint: GetAPISchema<
     // Add an item to the wishlist
     if (req.method === 'POST') {
-      const body = { ...req.body, customerToken }
+      const body = { ...req.body.variables, customerToken }
```
I think this is a breaking change for other providers. Please revert.

dominiksipowicz commented on site/.env.template:
```diff
@@ -24,6 +24,7 @@ BIGCOMMERCE_STORE_API_CLIENT_SECRET=
 
 NEXT_PUBLIC_SHOPIFY_STORE_DOMAIN=
 NEXT_PUBLIC_SHOPIFY_STOREFRONT_ACCESS_TOKEN=
+NEXT_PUBLIC_SHOPIFY_ADMIN_ACCESS_TOKEN=
```
same here, we don't want admin token to be public

chemmanouilidis replied:
same typo again, to be corrected.

dominiksipowicz commented on site/components/wishlist/WishlistCard/WishlistCard.tsx:
```diff
@@ -40,7 +42,7 @@ const WishlistCard: React.FC<{
       // If this action succeeds then there's no need to do `setRemoving(true)`
       // because the component will be removed from the view
-      await removeItem({ id: item.id! })
+      await removeItem({ productId: item.id, variantId: variant })
```
why not just keep the current API interface and just remove product with all variants of the same product? This will not be a breaking change

chemmanouilidis replied:
If you think this is a breaking change or will fail on other providers (because of the way I designed it) this is not a problem we can modify it to remove every item in wishlist and all its variants.

dominiksipowicz reviewed (changes requested):
Thank you for your contribution. It looks awesome and is a lot of work. Please see my comments and requests for changes. 

*Additional notes*:
 - we use `yarn` for this project, please remove `package-lock.json`
 - the main "local" build is failing

dominiksipowicz commented on site/components/product/ProductCard/ProductCard.tsx:
```diff
@@ -105,7 +105,7 @@ const ProductCard: FC<Props> = ({
                 className={s.wishlistButton}
                 productId={product.id}
-                variant={product.variants[0] as any}
+                variant={product.variants[0]}
```
Awesome!

chemmanouilidis commented on site/commerce.config.json:
```diff
@@ -2,8 +2,8 @@
     "search": true,
-    "wishlist": false,
-    "customerAuth": false,
+    "wishlist": true,
```
ok

chemmanouilidis commented on packages/commerce/src/api/endpoints/wishlist.ts:
```diff
@@ -34,13 +34,13 @@ const wishlistEndpoint: GetAPISchema<
     // Remove an item from the wishlist
     if (req.method === 'DELETE') {
-      const body = { ...req.body, customerToken }
+      const body = { ...req.body.variables, customerToken }
```
Ok, same as above, I will change the destructuring method.

chemmanouilidis commented:
Corrections Commit includes the following changes: 

**### Commerce API**

1. reverted commerce/api/endpoints/wishlist & commerce/types/wishlist  to source implementation to avoid problems with other providers and major breaking changes

**### Shopify provider** 

1. Modified shopify/api/endpoints/wishlist/add-item property destructuring to be compatible with the types used, 
2. remove-item also has the same destructuring method plus new filtering method that correctly filters the wishlist by productId AND variantID
3. Modified shopify/types/wishlist. RemoveItemHook type now correctly extends the core type from commerce/types/wishlist  by adding an optional property itemVariantId, modified WishlistSchema type to be compatible with the new functions used in wishlist/endpoints (add-item, remove-item etc) by adding variables type property for the body.
4. minor changes to use-add-item specifically in the way properties are being handed to the fetcher.
5. use-remove-item hook now complies with the new types declared passing through itemId and itemVariantId
6. modified env.template and const files to "hide" admin access token variable.

### **Site Components**

1. Modified WishlistButton removeItem function to be compatible across all providers
2. Modified WishlistCard removeItem function (same as above) and also added variant options labels for each product in your wishlist. This all summarizes my goal from the begining being a user should be able to add more variants of the same product in his wishlist and not just the main product. One, can desire both the blue and the red version of that T-Shirt that is currently out of stock in those colors.
3. Wishlist page midifications are mainly reverts to previous changes I have made. Now wishlist page should be the same as origin.

chemmanouilidis commented:
> Thank you for your contribution. It looks awesome and is a lot of work. Please see my comments and requests for changes.
> 
> _Additional notes_:
> 
> * we use `yarn` for this project, please remove `package-lock.json`
> * the main "local" build is failing

package lock removed.
local build was failing to me because it could not resolve the wishlist hooks from the "local" framework. wishlist folder of local package didn't have an index file with default exports. Should I add it and commit? adding the file to me, solves the problem.

chemmanouilidis commented:
now the main "local" build succeeds.

thearegee commented:
@chemmanouilidis thanks for this PR. Is there any progress here, did you ever use this in production?

chemmanouilidis commented:
> @chemmanouilidis thanks for this PR. Is there any progress here, did you ever use this in production?

I don't think any further progress is required. I currently use it in production and it seems stable.

curiouscrusher commented on packages/shopify/src/api/endpoints/wishlist/add-item.ts:
```diff
@@ -0,0 +1,84 @@
+      })!
+      metafieldInput = { id: wishlist?.id! }
+    } else {
+      metafieldInput = { namespace: 'my_fields', key: 'wishlist' }
```
@chemmanouilidis just a heads up, I believe you need a `type: 'json'` on this `metafieldInput`, otherwise the initial metafield API call will quietly fail and the Wishlist is never created.

leerob commented:
Hey there! Thank you for your contribution. We have decided to take Next.js Commerce in a new direction and will be closing out current PRs and issues due to this change. Please see this PR for more details: https://github.com/vercel/commerce/pull/966
